File: backend/score.py
from dataclasses import dataclass
from typing import Literal

Decision = Literal["TAKE", "MAYBE", "SKIP"]


@dataclass(frozen=True)
class Leg:
    distance_km: float
    duration_min: float


@dataclass(frozen=True)
class ScoreConfig:
    cost_per_km: float
    platform_commission: float
    min_ars_per_km: float
    min_ars_per_hr: float
    max_deadhead_ratio: float


@dataclass(frozen=True)
class Verdict:
    decision: Decision
    gross_ars: float
    net_ars: float
    profit_ars: float
    total_km: float
    total_min: float
    deadhead_km: float
    deadhead_ratio: float
    ars_per_km: float
    ars_per_hr: float
    reasons: list[str]


def compute_verdict(
    offer_ars: float,
    deadhead: Leg,
    trip: Leg,
    cfg: ScoreConfig,
) -> Verdict:
    total_km = deadhead.distance_km + trip.distance_km
    total_min = deadhead.duration_min + trip.duration_min
    deadhead_ratio = deadhead.distance_km / total_km if total_km > 0 else 0.0

    gross = offer_ars
    net = gross * (1.0 - cfg.platform_commission)
    cost = total_km * cfg.cost_per_km
    profit = net - cost

    ars_per_km = profit / total_km if total_km > 0 else 0.0
    ars_per_hr = profit / (total_min / 60.0) if total_min > 0 else 0.0

    reasons: list[str] = []
    if ars_per_km < cfg.min_ars_per_km:
        reasons.append(f"ARS/km {ars_per_km:.0f} < min {cfg.min_ars_per_km:.0f}")
    if ars_per_hr < cfg.min_ars_per_hr:
        reasons.append(f"ARS/hr {ars_per_hr:.0f} < min {cfg.min_ars_per_hr:.0f}")
    if deadhead_ratio > cfg.max_deadhead_ratio:
        reasons.append(
            f"deadhead {deadhead_ratio:.0%} > max {cfg.max_deadhead_ratio:.0%}"
        )

    if reasons:
        decision: Decision = "SKIP"
    elif (
        ars_per_km > 1.2 * cfg.min_ars_per_km
        and ars_per_hr > 1.2 * cfg.min_ars_per_hr
    ):
        decision = "TAKE"
    else:
        decision = "MAYBE"

    return Verdict(
        decision=decision,
        gross_ars=gross,
        net_ars=net,
        profit_ars=profit,
        total_km=total_km,
        total_min=total_min,
        deadhead_km=deadhead.distance_km,
        deadhead_ratio=deadhead_ratio,
        ars_per_km=ars_per_km,
        ars_per_hr=ars_per_hr,
        reasons=reasons,
    )

File: android/app/src/main/java/com/juanigsrz/driverhelper/Score.kt
package com.juanigsrz.driverhelper

import java.util.Locale
import kotlin.math.roundToInt

/**
 * On-device port of the backend `score.py` + the preferred (OCR-distance) path
 * of `main.py`. Field names on [VerdictOut] stay snake_case so [Notifier] reads
 * them unchanged from when the verdict came over the wire.
 */

data class Leg(val distanceKm: Double, val durationMin: Double)

data class ScoreConfig(
    val costPerKm: Double,
    val platformCommission: Double,
    val minArsPerKm: Double,
    val minArsPerHr: Double,
    val maxDeadheadRatio: Double,
) {
    companion object {
        /** Mirrors the backend CFG defaults in `main.py`. */
        val DEFAULT = ScoreConfig(
            costPerKm = 400.0,
            platformCommission = 0.25,
            minArsPerKm = 600.0,
            minArsPerHr = 8000.0,
            maxDeadheadRatio = 0.40,
        )
    }
}

data class VerdictOut(
    val decision: String,
    val gross_ars: Double,
    val net_ars: Double,
    val profit_ars: Double,
    val total_km: Double,
    val total_min: Double,
    val deadhead_km: Double,
    val deadhead_ratio: Double,
    val ars_per_km: Double,
    val ars_per_hr: Double,
    val reasons: List<String>,
)

private fun Double.fmt0(): String = String.format(Locale.US, "%.0f", this)

fun computeVerdict(
    offerArs: Double,
    deadhead: Leg,
    trip: Leg,
    cfg: ScoreConfig,
): VerdictOut {
    val totalKm = deadhead.distanceKm + trip.distanceKm
    val totalMin = deadhead.durationMin + trip.durationMin
    val deadheadRatio = if (totalKm > 0) deadhead.distanceKm / totalKm else 0.0

    val gross = offerArs
    val net = gross * (1.0 - cfg.platformCommission)
    val cost = totalKm * cfg.costPerKm
    val profit = net - cost

    val arsPerKm = if (totalKm > 0) profit / totalKm else 0.0
    val arsPerHr = if (totalMin > 0) profit / (totalMin / 60.0) else 0.0

    val reasons = mutableListOf<String>()
    if (arsPerKm < cfg.minArsPerKm) {
        reasons.add("ARS/km ${arsPerKm.fmt0()} < min ${cfg.minArsPerKm.fmt0()}")
    }
    if (arsPerHr < cfg.minArsPerHr) {
        reasons.add("ARS/hr ${arsPerHr.fmt0()} < min ${cfg.minArsPerHr.fmt0()}")
    }
    if (deadheadRatio > cfg.maxDeadheadRatio) {
        reasons.add(
            "deadhead ${(deadheadRatio * 100).roundToInt()}% > " +
                "max ${(cfg.maxDeadheadRatio * 100).roundToInt()}%"
        )
    }

    val decision = when {
        reasons.isNotEmpty() -> "SKIP"
        arsPerKm > 1.2 * cfg.minArsPerKm && arsPerHr > 1.2 * cfg.minArsPerHr -> "TAKE"
        else -> "MAYBE"
    }

    return VerdictOut(
        decision = decision,
        gross_ars = gross,
        net_ars = net,
        profit_ars = profit,
        total_km = totalKm,
        total_min = totalMin,
        deadhead_km = deadhead.distanceKm,
        deadhead_ratio = deadheadRatio,
        ars_per_km = arsPerKm,
        ars_per_hr = arsPerHr,
        reasons = reasons,
    )
}

/**
 * Mirrors `main.py:144-154`: first leg is the deadhead to pickup, second is the
 * trip. A single leg means the card only printed the trip itself. Returns null
 * when the offer can't be scored (no usable price, or no distance row parsed).
 */
fun evaluateOffer(price: Double?, legs: List<Leg>, cfg: ScoreConfig): VerdictOut? {
    if (price == null || price <= 0) return null
    if (legs.isEmpty()) return null
    val (deadhead, trip) = if (legs.size == 1) {
        Leg(0.0, 0.0) to legs[0]
    } else {
        legs[0] to legs[1]
    }
    return computeVerdict(price, deadhead, trip, cfg)
}
